### PR TITLE
Use c# back references for regex matching, refactor to rewrite/redirect actions

### PR DESCRIPTION
**Pull request by ZestyBread**
@Tratcher @natemcmaster @davidfowl

### CODE REVIEW

**Comment by ZestyBread** [comment]
The diff got enlarged due to file renaming, effectively what I'm doing is using the mod_rewrite implementation with default parameters for UrlRewrite to use.

File: src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RewriteAction.cs [hunk 1]
```diff
@@ -38,10 +62,25 @@ public override void ApplyAction(RewriteContext context, MatchResults ruleMatch,
                 FragmentString fragment;
                 UriHelper.FromAbsolute(pattern, out scheme, out host, out path, out query, out fragment);
 
+                if (query.HasValue)
+                {
+                    if (QueryStringAppend)
+                    {
+                        context.HttpContext.Request.QueryString = context.HttpContext.Request.QueryString.Add(query);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RewriteAction.cs]: extract `context.HttpContext.Request` to local variable

File: src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RedirectAction.cs [hunk 2]
```diff
@@ -2,49 +2,102 @@
...
-                context.HttpContext.Response.Headers[HeaderNames.Location] = pattern;
+                // If the request url has a query string and the target does not, append the query string
+                // by default.
+                if (QueryStringDelete)
+                {
+                    context.HttpContext.Response.Headers[HeaderNames.Location] = pattern;
+                }
+                else
+                {
+                    context.HttpContext.Response.Headers[HeaderNames.Location] = pattern + context.HttpContext.Request.QueryString;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RedirectAction.cs]: extract `context.HttpContext.Response` to a local variable

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RedirectAction.cs]: in all of these modified files

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RedirectAction.cs]: or consider adding helpers like these: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication/Events/BaseContext.cs#L17-L25

**Comment by Tratcher** [comment]
:shipit:

**Pull request merged**